common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchLogStore.scala

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool.logging
+
+import java.nio.file.{Path, Paths}
+
+import akka.actor.ActorSystem
+import akka.stream.scaladsl.Flow
+import akka.http.scaladsl.model._
+
+import whisk.core.entity.{ActivationLogs, Identity, WhiskActivation}
+import whisk.core.containerpool.logging.ElasticSearchJsonProtocol._
+import whisk.core.ConfigKeys
+
+import scala.concurrent.{Future, Promise}
+import scala.util.Try
+
+import spray.json._
+
+import pureconfig._
+
+case class ElasticSearchLogStoreConfig(protocol: String,
+                                       host: String,
+                                       port: Int,
+                                       path: String,
+                                       logMessageField: String,
+                                       activationIdField: String,
+                                       streamField: String,
+                                       actionField: String,
+                                       requiredHeaders: String = "")
+
+/**
+ * ElasticSearch based implementation of a DockerToActivationFileLogStore. When using the JSON log driver, docker writes
+ * stdout/stderr to JSON formatted files. Those files can be processed by a backend service asynchronously to store
+ * user logs in ElasticSearch. This log store allows user logs then to be fetched from ElasticSearch.
+ */
+class ElasticSearchLogStore(
+  system: ActorSystem,
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  destinationDirectory: Path = Paths.get("logs"),
+  elasticSearchConfig: ElasticSearchLogStoreConfig =
+    loadConfigOrThrow[ElasticSearchLogStoreConfig](ConfigKeys.elasticSearch))
+    extends DockerToActivationFileLogStore(system, destinationDirectory) {
+
+  // Schema of logs in ES
+  case class UserLogEntry(message: String, stream: String, time: String, action: String)
+
+  object UserLogEntry extends DefaultJsonProtocol {
+    implicit val serdes =
+      jsonFormat(
+        UserLogEntry.apply,
+        "message",
+        elasticSearchConfig.streamField,
+        "time_date",
+        elasticSearchConfig.actionField)
+  }
+
+  implicit val actorSystem = system
+
+  private val esClient = new ElasticSearchRestClient(
+    elasticSearchConfig.protocol,
+    elasticSearchConfig.host,
+    elasticSearchConfig.port,
+    httpFlow)
+  private val requiredHeaders = elasticSearchConfig.requiredHeaders match {
+    case headers if !headers.isEmpty => headers.split(",")
+    case _                           => Array.empty[String]
+  }
+  private val logQuery =
+    s"_type: ${elasticSearchConfig.logMessageField} AND ${elasticSearchConfig.activationIdField}: %s"
```

Comment:
Have you thought about making this a def taking the additional parameter? This way, it is an implementation detail how the additional %s is passed in. You shouldn't need that contextual knowledge on the caller side.